Add `label` filter for `docker system prune`

yongtang opened the pull request:
**- What I did**

This fix tries to address the issue raised in #29999 where it was not possible to mask these items (like important non-removable stuff) from `docker system prune`.

**- How I did it**


This fix adds `label` and `label!` field for `--filter` in `system prune`, so that it is possible to selectively prune items like:
```
$ docker container prune --filter label=foo

$ docker container prune --filter label!=bar
```

**- How to verify it**

Additional unit tests and integration tests have been added.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

![kittens27](https://cloud.githubusercontent.com/assets/6932348/22620690/08cffffc-eac6-11e6-9c38-97fc9b0e8969.jpg)


~This fix fixes~
This fix is an improvement for #29999.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add `label` filter for `docker system prune`

This fix tries to address the issue raised in 29999 where it was not
possible to mask these items (like important non-removable stuff)
from `docker system prune`.

This fix adds `label` and `label!` field for `--filter` in `system prune`,
so that it is possible to selectively prune items like:
```
$ docker container prune --filter label=foo

$ docker container prune --filter label!=bar
```

Additional unit tests and integration tests have been added.

This fix fixes 29999.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
/cc @cpuguy83 @thaJeztah

cpuguy83 commented:
I wouldn't really call this fixing it since it's a pretty bad interface, and not really any better than manually doing `docker <object> rm $(docker <object> list --filter label=foo)`.

One potential solution to this would be to have the prune command read default filters from the cli config, but even then I'm not sure it's a particularly great interface... though definitely an improvement.

vdemeester commented:
Yeah I agree with @cpuguy83, improvement but not fixing it :angel:. Maybe we could also have a daemon configuration that defines some labels that should be protected by default ?

yongtang commented:
Thanks @cpuguy83 @vdemeester for the review. I updated the PR and now it is possible to specify `pruneFilters` in `config.json`. That should allow pre-configuring the labels for prune to include or exclude. Please take a look.

The docs will be updated if we are ok with the interface.

yongtang commented:
The PR description has also been updated to remove the `fixes`.

cpuguy83 commented:
Design OK, will need a rebase.

ping @vdemeester design ok for you?

vdemeester reviewed (commented):
Design LGTM to me
/cc @thaJeztah @mlaventure

mlaventure commented:
How does this behave if I have `label!=ignore` in the config file but I specify `label=ignore` in the cli?  :)

yongtang commented:
The PR has been rebased.

yongtang commented:
@mlaventure In case of conflicting filtering, the current behavior is to ignore (not prune). I added a test case to show the behavior.

Not sure the best way to handle this scenario so let me know if we want to have special handling (like throw out an error or warning?).

mlaventure commented:
Hum, that's a bit annoying a behavior IMHO.

I think we usually have the cli flags supersede the values from the config. 

If it doesn't in that case, I'd have to either update temporarily my config file or resort to cascading docker commands.

@cpuguy83 @vdemeester WDYGT?

cpuguy83 commented:
Agree it should supersede.

Once "--dry-run" it will help in such cases as well to actually see what would happen.

cpuguy83 commented:
*is implemented.

vdemeester commented:
Agree with @cpuguy83 @mlaventure, the cli argument should superseed the config 👼

yongtang commented:
Thanks all for the review. The PR has been updated and now CLI filter will supersede the config.json filter.

cpuguy83 commented:
@mlaventure Good to move to code-review?

mlaventure commented:
Yep!

cpuguy83 commented on cli/command/volume/prune.go:
```diff
@@ -6,13 +6,15 @@ import (
 type pruneOptions struct {
-	force bool
+	force  bool
+	filter opts.FilterOpt
```
This is not properly initialized, getting a panic when trying to add filters.

yongtang replied:
Thanks @cpuguy83 for the review. The panic has been fixed.

cpuguy83 commented on integration-cli/docker_cli_prune_unix_test.go:
```diff
@@ -111,3 +114,61 @@ func (s *DockerSuite) TestPruneContainerUntil(c *check.C) {
 	c.Assert(strings.TrimSpace(out), checker.Contains, id2)
 }
+
+func (s *DockerSuite) TestPruneContainerLabel(c *check.C) {
```
Looks like we need to test each of the subcommands

yongtang replied:
Thanks. Additional tests have been added.

thaJeztah commented:
ping @yongtang I see @cpuguy83 left some review notes :smile:

yongtang commented:
Thanks all for the review. The PR has been updated.

cpuguy83 reviewed (approved):
LGTM

mlaventure commented on integration-cli/docker_cli_prune_unix_test.go:
```diff
@@ -111,3 +114,177 @@ func (s *DockerSuite) TestPruneContainerUntil(c *check.C) {
+	c.Assert(strings.TrimSpace(out), checker.Not(checker.Contains), id1)
+	c.Assert(strings.TrimSpace(out), checker.Not(checker.Contains), id2)
+	c.Assert(strings.TrimSpace(out), checker.Not(checker.Contains), id3)
+	c.Assert(strings.TrimSpace(out), checker.Contains, id4)
```
`checker.Equals`  on `id4` would be enough in that case

mlaventure commented on integration-cli/docker_cli_prune_unix_test.go:
```diff
@@ -111,3 +114,177 @@ func (s *DockerSuite) TestPruneContainerUntil(c *check.C) {
+	c.Assert(strings.TrimSpace(out), checker.Contains, id4)
+
+	out, _ = dockerCmd(c, "container", "prune", "--force", "--filter", "label=foo")
+	c.Assert(strings.TrimSpace(out), checker.Contains, id1)
```
`checker.Equals`  on `id1` would be enough in that case (similar comment for the other subcommands)

mlaventure reviewed (approved):
nit on checker use, but LGTM

yongtang commented:
Thanks @mlaventure for the review. Currently the output of `prune` also contains `Deleted Images` and `Total reclaimed space:` so it may not be possible to use `Equals`.


```
ubuntu@ubuntu:~/docker$ docker image prune -f
Deleted Images:
deleted: sha256:9485e5f3b535139c864b249fc8b135d1dc3e2907f4b0cfda6012b9cc045c8b53
deleted: sha256:483ef533be5120f014ec2d9298b0daa240ec7f0da3b2fa5d4e4bdcd36144491b
deleted: sha256:ca96556fd1b6ea07ac0e9f1f7d9b4d7440780752484e620199a53a91b36b1731
deleted: sha256:725617c4f06697290b288a91ebd884b1f2a1d9e6da3e7514d31c2bab09e65d73
deleted: sha256:85b88b89c53f89f7f4e93941882c751e2af8b25477ad0b23629663014bcbf084
deleted: sha256:b92809f737413ddbd4efecfbf570045f177376a3f4992aa4b9a7ef52e7085c69
deleted: sha256:80ec69f46395bfcbda5de6c2a84f6a5c5a32388b2f7fe59dcb3b1463a45b7779
deleted: sha256:4a389c79f1a7e8d2e5be44e2c2c84e5685a113b80f2a2deac3206c348eaacea4
deleted: sha256:cb9343a57b9a4745c75b1b37607dfce3e9152802e54ba22af955b387d42e7de7
deleted: sha256:8b85a29f5f6e4198db2d4cdd836da03e456197031528daacd8f7be0093836309
deleted: sha256:ceeb09215557367d4781c899034b912381c047bb2899ea3cb3e28fafeda82235
deleted: sha256:0c7f605ab25bd85021421084e557bae9047ba37733ee2a45e988525d18085944

Total reclaimed space: 195.1 MB
```

I am thinking it might make sense to add an additional flag of `-q` to `image prune` (and others) so that only the digests of images are printed?

vdemeester reviewed (approved):
LGTM 🐸
The `-q` flags can/should be handled in a follow-up PR 😉 
Moving to docs-review, cc @thaJeztah

thaJeztah commented:
@yongtang can you update the;

- system prune CLI reference; https://github.com/docker/docker/blob/master/docs/reference/commandline/system_prune.md
- network prune CLI reference; https://github.com/docker/docker/blob/master/docs/reference/commandline/network_prune.md
- container prune CLI reference; https://github.com/docker/docker/blob/master/docs/reference/commandline/container_prune.md
- image prune CLI reference; https://github.com/docker/docker/blob/master/docs/reference/commandline/image_prune.md
- volume prune CLI reference; https://github.com/docker/docker/blob/master/docs/reference/commandline/volume_prune.md
- the API history; https://github.com/docker/docker/blob/master/docs/api/version-history.md
- the swagger.yml https://github.com/docker/docker/blob/master/api/swagger.yaml

Given that 17.05 code freeze is today; let me know if you want to handle the docs in a follow up

vdemeester commented:
I am all for doing a follow-up for docs, let's get this merge.
@yongtang ping me if you want us to do the follow-up 😉

yongtang commented:
Thanks @vdemeester. I am currently updating the docs. Will try to create a PR either late today or early tomorrow. Hopefully I could have the doc PR ready shortly today.
